crates/ra_parser/src/grammar/items/use_item.rs

```diff
@@ -14,7 +14,7 @@ pub(super) fn use_item(p: &mut Parser, m: Marker) {
 /// Note that this is called both by `use_item` and `use_tree_list`,
 /// so handles both `some::path::{inner::path}` and `inner::path` in
 /// `use some::path::{inner::path};`
-fn use_tree(p: &mut Parser, top_level: bool) {
```

Comment:
This substantially reverts #1866; The new behavior also fixes the infinite loop and results in a better error-recovery on the provided test case.